src/lib/io/sol-ipm.c

```diff
@@ -0,0 +1,323 @@
+
+    SOL_VECTOR_FOREACH_IDX (&ipm_receivers, receiver, idx) {
+        if (receiver->id == id) {
+            struct sol_blob_ipm *remote_blob;
+
+            if (!receiver->last_remote_blob) {
+                SOL_WRN("Remote blob for IPM id %" PRIu32 " disappered!", id);
+                return;
+            }
+
+            remote_blob = calloc(1, sizeof(struct sol_blob_ipm));
+            if (!remote_blob) {
+                SOL_WRN("Could not create blob to deliver IPM message");
+                receiver->last_remote_blob = NULL;
+                return;
+            }
+
+            sol_blob_setup((struct sol_blob *)remote_blob, &SOL_BLOB_TYPE_IPM,
+                receiver->last_remote_blob->mem,
+                receiver->last_remote_blob->size);
+
+            remote_blob->remote = receiver->last_remote_blob;
+            receiver->last_remote_blob = NULL;
+            receiver->receive_cb((void *)receiver->data, id,
+                (struct sol_blob *)remote_blob);
+            return;
+        }
+    }
+
+    SOL_WRN("Processed IPM id %" PRIu32 " but no receiver was found!", id);
+}
+
+static void
+ipm_reaper_process(void *data)
+{
+    int idx;
+    struct sol_ipm_sent_message *sent;
+    struct sol_blob *message = data;
+
+    SOL_VECTOR_FOREACH_IDX (&ipm_sent_messages, sent, idx) {
+        if (sent->message == message) {
+            if (sent->message_consumed_cb)
+                sent->message_consumed_cb((void *)sent->data, sent->id, message);
+            sol_blob_unref(message);
+            sol_vector_del(&ipm_sent_messages, idx);
+            return;
+        }
+    }
+}
+
+/* Note: this function runs in interrupt context */
+static void
+ipm_receiver_cb(void *context, uint32_t id, volatile void *data)
+{
+    int idx;
+    struct sol_blob *remote_blob = (struct sol_blob *)AS_PTR(data);
+    struct sol_ipm_receiver_entry *receiver;
+    struct mainloop_event me;
+
+    /* id == 0 messages. Those messages signal sender
+    * core that blob was consumed and may be freed */
+    if (id == SOL_IPM_ID_FREE_REMOTE) {
+        me.cb = ipm_reaper_process;
+        me.data = remote_blob;
+        sol_mainloop_event_post(&me);
+        return;
+    }
+
+    //TODO rethink this. What if we receive another message with same id
+    //before processing current one?
+    me.cb = ipm_receiver_process,
+    me.data = (void *)id; /* I could send `receiver` entry, and avoid yet
+                             another loop to find receiver. Although currently,
+                             Zephyr implementation does process events before
+                             timeouts, I'm not sure we can count on that
+                             `ipm_vectors` is not going to be changed in the
+                             meanwhile.*/
+
+    SOL_VECTOR_FOREACH_IDX (&ipm_receivers, receiver, idx) {
+        if (receiver->id == id) {
+            receiver->last_remote_blob = remote_blob;
+            int r = sol_mainloop_event_post(&me);
```

Comment:
On 05-05-2016 18:53, Gustavo Sverzut Barbieri wrote:
> In src/lib/io/sol-ipm.c
> <https://github.com/solettaproject/soletta/pull/1981#discussion_r62261797>:
>
>> +    }
>> +
>> +    //TODO rethink this. What if we receive another message with same id
>> +    //before processing current one?
>> +    me.cb = ipm_receiver_process,
>> +    me.data = (void *)id; /* I could send `receiver` entry, and avoid yet
>> +                             another loop to find receiver. Although currently,
>> +                             Zephyr implementation does process events before
>> +                             timeouts, I'm not sure we can count on that
>> +                             `ipm_vectors` is not going to be changed in the
>> +                             meanwhile.*/
>> +
>> +    SOL_VECTOR_FOREACH_IDX (&ipm_receivers, receiver, idx) {
>> +        if (receiver->id == id) {
>> +            receiver->last_remote_blob = remote_blob;
>> +            int r = sol_mainloop_event_post(&me);
>
> I doubt this is passing check-style.py...

Well, it is... but again, debug leftover

>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/solettaproject/soletta/pull/1981/files/9efa022b5f0960e84101624c9866a4cd1a343cbb#r62261797>
>